<a href="https://colab.research.google.com/github/mlepinski/Python-Worksheets/blob/master/Week_12_Worksheet.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**SPREADSHEET DATA**

Spreadsheet data -- data with rows and columns -- comes up all the time in the natural and social sciences. 

This data is very commonly stored in CSV files (which stands for Comma Separated Values). If you are used to using a tool like Google Sheets, you can check the File menu and see that Google Sheets has an option to Download the data as a CSV file. Microsoft Excel and similar programs also have the option to save as a CSV file. Also, if you find interesting data on the Internet, it is very likely to be available as a CSV file.

So a CSV file is just a file, and so we can use the techniques from the last assignment to process CSV files just like we did for the file of Shakespeare's Sonnets. 

For example, I put on Github some data on the area and population of US states. The following code -- similar to what we did on the last assignment -- can be used to see that there are two states that include the word "North". 



In [ ]:
import urllib.request

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
my_file = urllib.request.urlopen(file_name)

long_string = my_file.read()
long_string = long_string.decode("utf-8")

no_line_breaks = long_string.replace("\n", " ")
no_commas = no_line_breaks.replace(",", " ")
lower_case = no_commas.lower()

list_of_words = lower_case.split()

word_count = 0 

for word in list_of_words:
  if word == "north":
    word_count = word_count + 1

print("The word North appears", word_count, "times in the file")

However, if you want to do more complicated things with your data, this approach gets messy fast. 

Therefore, in practice, when Python folks want to work with spreadsheet data they use a special Python package, called Pandas. Pandas is specially designed to make it easier to work with CSV files and it is one of the most commonly used Python packages. 

The example code below uses the Pandas read_csv command to grab a CSV file off the Internet. Pandas allows you to store the data from a CSV file in a Python variable. We call the resulting variable a "Pandas dataframe"

Pandas provides a bunch of useful commands for working with dataframes. They may be a bit confusing at first, but they are powerful once you get used to them. 

Below we use the comamnd **head()** to print out the first few rows of a dataframe. Since dataframes are sometimes very large, it is often useful to just print out the first few rows to get a since of what the data looks like

In [ ]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

state_data.head()

Name State    Area       Pop
0     Alabama    AL   52423   4040587
1      Alaska    AK  656424    550043
2     Arizona    AZ  114006   3665228
3    Arkansas    AR   53182   2350725
4  California    CA  163707  29769021

Pandas uses square brackets to identify particular columns of data. 

So for example, we could print the column of State Populations as follows:



In [ ]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

print(state_data["Pop"])


I wonder what the largest state population is? 

We could use a for loop to loop through the population data and figure that out.

In [ ]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

biggest_pop = 0

for pop in state_data["Pop"]:
  if pop > biggest_pop:
    biggest_pop = pop

print(biggest_pop)

29769021


Wow! One state has almost 30 Million people. I wonder what state that is?

(Also, this might be a good time to mention that this data is a bit old. It is from a previous census. When we get the new data from the 2020 census, I will need to update my Github repository)

In order to work with both the Name and Population of a state, we would need to loop through the rows in the data frame and not loop through one column. 

The command **iterrows()** (short for iterate the rows) allows us to loop through the rows of a data frame. 

The following code loops through the rows of a data frame, calling the row number **row_num** and the data for the row is **row**. 

As with any for loop, we can name the loop variables anything we want. The the thing that is different here (as compared to most loops that we do) is that **iterrows()** forces us to have two loop variables (one for the row number and one for the data) instead of one. In this example, we give a name to the row number (because that is what **iterrows()** wants us to do) but then we ignore the row number inside of our loop.

In [ ]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

biggest_pop = 0

for (row_num, row)  in state_data.iterrows():
  print(row["Name"], row["Pop"])



So we successfully looped through the Rows of our data and printed out the state name and the population. 

That is a good start, now we just need to keep track of the name and population for the biggest state we have seen so far. 


In [ ]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

biggest_pop = 0
biggest_name = " "

for (row_num, row)  in state_data.iterrows():
  if row["Pop"] > biggest_pop:
    biggest_pop = row["Pop"]
    biggest_name = row["Name"]

print(biggest_name, biggest_pop)



---



---



---
**SELECTING ROWS WE CARE ABOUT**

Pandas also gives you the ability to easily grab the row (or rows) that satisfy a particular condition. 

For example, let's say I want to find the row with the State (abbreviation) FL


In [ ]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

fl = state_data[ state_data["State"]== "FL"]

print(fl)

      Name State   Area       Pop
9  Florida    FL  65756  12937926


This gives an alternative (easier) way to find the sate with the largest population. (That **iterrows** stuff was kind of weird, I like this way better)

First, we use the built-in **max** function to find the largest population (without looping!) 

In [ ]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

max_pop = state_data["Pop"].max()

print(max_pop)

Pandas actually has a bunch of useful built-in functions that can often give us the answer we are looking for without needing to write a loop.

For example, instead of **state_data["Pop"].max()** we could have asked for the mean (average) with **state_data["Pop"].mean()** or the median population with **state_data["Pop"].median()**

Now let's get back to our example for finding the largest population state. Once we know the population of the largest state, it is really easy to pick out the state that has this particular population ... This is just like when we asked Pandas to find the row where the abbreviation was "FL" ... Except this time we want the row where the population is max_pop (29769021).

In [10]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

max_pop = state_data["Pop"].max()

max_pop_row = state_data[ state_data["Pop"]== max_pop]

print(max_pop_row)

         Name State    Area       Pop
4  California    CA  163707  29769021




---



---



---

**ADDING A NEW COLUMN**

So the Area column in our data uses the units of Square Miles. Suppose we wanted the Area in Square Kilometers (metric system) instead od Square Miles. 

**Note:** This type of unit conversation comes up all the time in scientific data. 

We can convert from Square Miles to Square Kilometers by multiplying the area by a conversation factor of 2.59 (one square mile is 2.59 square kilometers). 

Here is how we add a new column to our existing data table. 


In [12]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

state_data["Area Kilometers"] = state_data["Area"] * 2.59

state_data.head()

Name State    Area       Pop  Area Kilometers
0     Alabama    AL   52423   4040587        135775.57
1      Alaska    AK  656424    550043       1700138.16
2     Arizona    AZ  114006   3665228        295275.54
3    Arkansas    AR   53182   2350725        137741.38
4  California    CA  163707  29769021        424001.13

In the above example, I chose to name the new column "Area Kilometers". However, I could have named it anything that I wanted by just putting a different name inside the square brackets. 

As a final example, let's figure out what the average area of a state is in Square Kilometers.



In [ ]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

state_data["Area Kilometers"] = state_data["Area"] * 2.59

average_km_area = state_data["Area Kilometers"].mean()

print(average_km_area)

Okay, now let's figure out which state is closest to the average area. 

I am going to loop through the areas of each state and keep track of the smallest difference (gap/distance) between the state's area and the average area. 

**Note:** I want the difference between the current state's area and the average area to be positive, so I use the built in **abs** (absolute value) function which just makes a number positive by getting rid of the negative sign if the number is less than zero. 

In [ ]:
import pandas

file_name = "https://raw.githubusercontent.com/mlepinski/Python-Worksheets/master/states.csv"
state_data = pandas.read_csv(file_name)

state_data["Area Kilometers"] = state_data["Area"] * 2.59

average_km_area = state_data["Area Kilometers"].mean()

state_data["Area Difference"] = abs( state_data["Area Kilometers"] - average_km_area)

smallest_difference = state_data["Area Difference"].min()

close_to_average_row = state_data[ state_data["Area Difference"] == smallest_difference]

print("This state is closest to the average area of a state:")
print(close_to_average_row)